### PR TITLE
[ci] enforce flake8 (fixes #161)

### LINKED ISSUE #161: [ci] add flake8
**Is your feature request related to a problem? Please describe.**

The project does not currently have any automatic protections against some classes of issue that can be detected by `flake8`, including:

* unused imports
* unused variables
* duplicated test names (which leads to  `pytest` skipping tests)
* f-strings used on strings with no templating

**Describe the solution you'd like**

I believe this project should use `flake8` for linting, and should store the configuration for that tool in a `setup.cfg` file (https://flake8.pycqa.org/en/latest/user/configuration.html).

Adding this tool to the project's testing setup would reduce the effort required for pull request authors and reviewers to detect such issues, and would reduce the risk of code changes with such issues making it to `main`.

**Describe alternatives you've considered**

N/A

**Additional context**

Here is the result of running `flake8` (ignoring style-only issues) on the current latest commit on `main`.

```shell
flake8 \
    --ignore=E126,E128,E203,E241,E261,E302,E303,E402,E501,W503,W504,W605 \
    .
```

```text
./hamilton/function_modifiers.py:50:13: F402 import 'node' from line 11 shadowed by loop variable
./hamilton/function_modifiers.py:169:30: F541 f-string is missing placeholders
./hamilton/function_modifiers.py:874:21: F541 f-string is missing placeholders
./hamilton/function_modifiers.py:878:34: F541 f-string is missing placeholders
./hamilton/graph.py:172:53: F821 undefined name 'graphviz'
./hamilton/graph.py:195:92: F821 undefined name 'networkx'
./hamilton/graph.py:316:13: F401 'graphviz' imported but unused
./hamilton/graph.py:451:17: F841 local variable 'e' is assigned to but never used
./hamilton/node.py:3:1: F401 'typing.Collection' imported but unused
./hamilton/driver.py:24:5: F811 redefinition of unused 'node' from line 13
./hamilton/data_quality/default_validators.py:227:30: F541 f-string is missing placeholders
./hamilton/data_quality/default_validators.py:394:9: F401 'pandera' imported but unused
./hamilton/data_quality/default_validators.py:396:21: F541 f-string is missing placeholders
./hamilton/data_quality/pandera_validators.py:1:1: F401 'typing.Any' imported but unused
./hamilton/data_quality/pandera_validators.py:21:16: F541 f-string is missing placeholders
./tests/test_function_modifiers.py:620:1: F811 redefinition of unused 'test_tags_invalid_value' from line 607
./tests/resources/bad_functions.py:5:1: F401 'tests.resources.only_import_me' imported but unused
./tests/resources/layered_decorators.py:1:1: F401 'hamilton.function_modifiers.tag' imported but unused
./tests/resources/cyclic_functions.py:5:1: F401 'tests.resources.only_import_me' imported but unused
./tests/resources/data_quality.py:1:1: F401 'numpy as np' imported but unused
./examples/ray/hello_world/run_rayworkflow.py:1:1: F401 'importlib' imported but unused
./graph_adapter_tests/h_ray/test_h_ray_workflow.py:1:1: F401 'tempfile' imported but unused
```

If maintainers are interested, I'd be happy to put up a pull request proposing this.

Thanks for your time and consideration.

### PR DESCRIPTION
**Pull request by jameslamb**
Proposes enforcing `flake8`'s checks on for pull requests to be merged.

In my experience, `flake8` is very helpful in detecting correctness and efficiency related issues such as:

* unused imports
* unused variables
* duplicated (and therefore ignored) unit tests functions
* unnecessary use of `f-strings`

And many more.

## Changes

"resolves" the following `flake8` warnings by annotating them with `#noqa` comments.

```text
./hamilton/graph.py:113:6: F821 undefined name 'graphviz'
./hamilton/graph.py:139:6: F821 undefined name 'networkx'
./hamilton/graph.py:275:13: F401 'graphviz' imported but unused
./hamilton/data_quality/default_validators.py:415:9: F401 'pandera' imported but unused
./tests/resources/bad_functions.py:5:1: F401 'tests.resources.only_import_me' imported but unused
./tests/resources/cyclic_functions.py:5:1: F401 'tests.resources.only_import_me' imported but unused
```

Adds the latest version of `flake8` as a new `pre-commit` task, ensuring that it's run in CI.

Adds a root-level `setup.cfg` containing configuration for `flake8`, including excluding some style-only warnings that are currently raised running `flake8 .` on `main`.

## Testing

```shell
pre-commit run --all-files
```

## Notes

*Why `setup.cfg` instead of just adding configuration in `.pre-commit-config.yaml`?*

Putting configuration for tools like `flake8` into the standard files used in Python projects (e.g. `setup.cfg` or `pyproject.toml`) decouples the configuration from the thing running the tool. e.g. if this project were to switch from `pre-commit run` to a `Makefile` or shell script for running these checks, none of the `flake8` configuration would need to be moved.

If maintainers feel strongly about not adding this file, I'd be happy to move this configuration back to `.pre-commit-config.yaml`.

Thanks for your time and consideration!

## Checklist

- [x] PR has an informative and human-readable title (this will be pulled into the release notes)
- [x] Changes are limited to a single goal (no scope creep)
- [x] Code can be automatically merged (no conflicts)
- [x] Code passed the pre-commit check & code is left cleaner/nicer than when first encountered.
- [ ] Passes all existing automated tests
- [x] Any _change_ in functionality is tested
- [x] New functions are documented (with a description, list of inputs, and expected output)
- [x] Placeholder code is flagged / future TODOs are captured in comments
- [x] Project documentation has been updated if adding/changing functionality.
- [ ] Reviewers requested with the [Reviewers tool](https://help.github.com/articles/requesting-a-pull-request-review/) :arrow_right:

### Python - local testing

- [ ] python 3.6
- [ ] python 3.7

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -26,3 +26,7 @@ repos:
     hooks:
     -   id: isort
         args: ["--profile", "black", --line-length=100]
+-   repo: https://github.com/pycqa/flake8
+    rev: 5.0.4
```
Comment (jameslamb) [inline .pre-commit-config.yaml:30]: this is the latest version as of today: https://pypi.org/project/flake8/#history

**Review by elijahbenizzy — APPROVED** [review]
LGTM!

**Comment by elijahbenizzy** [comment]
Happy with this, merging!

**Comment by jameslamb** [comment]
thanks @elijahbenizzy ! I think it'll be useful 😊

**Pull request merged**